### DISCUSSION TITLE
ERROR weewx.cheetahgenerator: Generate failed with exception '<type 'exceptions.UnicodeDecodeError'>'

### DISCUSSION THREAD

**Discussion by frahugras** [poblabs/weewx-belchertown#813] [Q&A] [answered]
Hello dears, i've installed another PSW system using WeeWX and Belchertown skin (amazing...); but this time i face this error and after several night time debugging hours no matter solve/understand it; can you please help? thanks! Fra

Aug 17 15:25:48 kittyhawk weewx[901] DEBUG weewx.reportengine: Running report 'StandardReport'
Aug 17 15:25:48 kittyhawk weewx[901] DEBUG weewx.reportengine: Found configuration file /etc/weewx/skins/Belchertown/skin.conf for report 'StandardReport'
Aug 17 15:25:48 kittyhawk weewx[901] DEBUG weewx.cheetahgenerator: Using search list ['weewx.cheetahgenerator.Almanac', 'weewx.cheetahgenerator.Station', 'weewx.cheetahgenerator.Current', 'weewx.cheetahgenerator.Stats', 'weewx.cheetahgenerator.UnitInfo', 'weewx.cheetahgenerator.Extras', 'weewx.cheetahgenerator.JSONHelpers', u'user.belchertown.getData']
Aug 17 15:25:48 kittyhawk weewx[901] DEBUG weewx.manager: Daily summary version is 4.0
Aug 17 15:25:49 kittyhawk weewx[901] INFO weewx.restx: PWSWeather: Published record 2022-08-17 15:25:00 CEST (1660742700)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: Generate failed with exception '<type 'exceptions.UnicodeDecodeError'>'
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: **** Ignoring template /etc/weewx/skins/Belchertown/json/weewx_data.json.tmpl
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: **** Reason: 'utf8' codec can't decode byte 0xec in position 16: unexpected end of data
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****  Traceback (most recent call last):
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/cheetahgenerator.py", line 326, in generate
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      unicode_string = compiled_template.respond()
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "_etc_weewx_skins_Belchertown_json_weewx_data_json_tmpl.py", line 2785, in respond
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/cheetahgenerator.py", line 691, in filter
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      filtered = six.text_type(val)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/tags.py", line 452, in __unicode__
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      return unicode(vh)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/units.py", line 1169, in __unicode__
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      return self.toString()
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/units.py", line 1159, in toString
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      localize=localize)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/units.py", line 840, in toString
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      s = self._to_string(val_t, context, addLabel, useThisFormat, None_string, localize)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/weewx/units.py", line 900, in _to_string
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      val_ustr = six.ensure_text(val_str)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/share/weewx/six.py", line 941, in ensure_text
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      return s.decode(encoding, errors)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****    File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****      return codecs.utf_8_decode(input, errors, True)
Aug 17 15:25:49 kittyhawk weewx[901] ERROR weewx.cheetahgenerator: ****  UnicodeDecodeError: 'utf8' codec can't decode byte 0xec in position 16: unexpected end of data
Aug 17 15:25:50 kittyhawk weewx[901] INFO weewx.cheetahgenerator: Generated 10 files for report StandardReport in 1.55 seconds
Aug 17 15:25:50 kittyhawk weewx[901] INFO weewx.reportengine: Copied 2 files to /var/www/html/weewx
Aug 17 15:25:50 kittyhawk weewx[901] DEBUG weewx.manager: Daily summary version is 4.0
Aug 17 15:25:50 kittyhawk weewx[901] DEBUG root: Start time is 1660687200 and end time is 1660773600
Aug 17 15:25:50 kittyhawk weewx[901] DEBUG root: Interval is: 300
Aug 17 15:25:53 kittyhawk weewx[901] DEBUG weewx.reportengine: Report 'Belchertown' not enabled. Skipping.
Aug 17 15:25:53 kittyhawk weewx[901] DEBUG weewx.reportengine: Report 'FTP' not enabled. Skipping.
Aug 17 15:25:53 kittyhawk weewx[901] DEBUG weewx.reportengine: Report 'RSYNC' not enabled. Skipping.

**Comment by norecords** [comment, marked as answer]
what version of Weewx and Belchertown do you use ?
I can see that you have python 2.7 in use, did you try with python 3 ?

**Reply by frahugras** [reply]
Hello, thanks for you reply; i'm sorry i should already inserted infos on first inquiy; please find below and i didn't try upgrade python environment




pi@kittyhawk:/etc/weewx/scripts$ sudo wee_debug --info 
Using verbosity=1, displaying most info

wee_debug output will be sent to stdout(console)

Using configuration file /etc/weewx/weewx.conf
Using database binding 'wx_binding', which is bound to database 'archive_sqlite'

System info
  Platform:       Linux-5.10.63-v7+-armv7l-with-debian-10.11
  Python Version: 2.7.16

Load Information
  1 minute load average:  0.25
  5 minute load average:  0.22
  15 minute load average: 0.15

General Weewx info
  Weewx version 4.5.1 detected.

Station info
  Station type: FineOffsetUSB
  Driver:       weewx.drivers.fousb

Driver info
[FineOffsetUSB]
    # This section is for the Fine Offset series of weather stations.
    
    # The station model, e.g., WH1080, WS1090, WS2080, WH3081
    model = WH1080
    
    # How often to poll the station for data, in seconds
    polling_interval = 60
    
    # The driver to use:
    driver = weewx.drivers.fousb

Currently installed extensions
Extension Name    Version   Description
Belchertown       1.2       A clean modern skin with real time streaming updates and interactive charts. Modeled after BelchertownWeather.com

Archive info
  Database name:        weewx.sdb
  Table name:           archive
  Version               4.0
  Unit system:          1 (US)
  First good timestamp: 2022-08-16 17:44:54 CEST (1660664694)
  Last good timestamp:  2022-08-20 14:20:00 CEST (1660998000)
  Number of records:    1643
  weewx (weewx.conf) is set to use an archive interval of 300 seconds.
  The station hardware was not interrogated in determining archive interval.




  kittyhawk weewx[11127] INFO __main__: Initializing weewx version 4.5.1
Aug 20 14:06:45 kittyhawk weewx[11127] INFO __main__: Using Python 2.7.16 (default, Oct 10 2019, 22:02:15) #012[GCC 8.3.0]
Aug 20 14:06:45 kittyhawk weewx[11127] INFO __main__: Platform Linux-5.10.63-v7+-armv7l-with-debian-10.11
Aug 20 14:06:45 kittyhawk weewx[11127] INFO __main__: Locale is 'it_IT'
Aug 20 14:06:45 kittyhawk weewx[11127] INFO __main__: Using configuration file /etc/weewx/weewx.conf
Aug 20 14:06:45 kittyhawk weewx[11127] INFO __main__: Debug is 1
Aug 20 14:06:45 kittyhawk weewx[11127] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
Aug 20 14:06:45 kittyhawk weewx[11127] INFO weewx.drivers.fousb: driver version is 1.20
Aug 20 14:06:45 kittyhawk weewx[11127] INFO weewx.drivers.fousb: polling mode is PERIODIC
Aug 20 14:06:45 kittyhawk weewx[11127] INFO weewx.drivers.fousb: polling interval is 60
Aug 20 14:06:45 kittyhawk weewx[11127] INFO weewx.drivers.fousb: found station on USB bus= device=
Aug 20 14:06:49 kittyhawk weewx[11143] INFO __main__: Initializing weewx version 4.5.1
Aug 20 14:06:49 kittyhawk weewx[11143] INFO __main__: Using Python 2.7.16 (default, Oct 10 2019, 22:02:15) #012[GCC 8.3.0]
Aug 20 14:06:49 kittyhawk weewx[11143] INFO __main__: Platform Linux-5.10.63-v7+-armv7l-with-debian-10.11
Aug 20 14:06:49 kittyhawk weewx[11143] INFO __main__: Locale is 'it_IT'

**Reply by norecords** [reply]
Try to set Python 3 with update-alternatives and reboot

**Reply by norecords** [reply]
You can follow this link https://linuxconfig.org/how-to-change-from-default-to-alternative-python-version-on-debian-linux

**Reply by frahugras** [reply]
Hello,

i’ve managed to follow suggestion; please check attached logs; unfortunately error is still there

I see

Aug 23 11:57:48 kittyhawk weewx[864] INFO __main__: Initializing weewx version 4.5.1
Aug 23 11:57:48 kittyhawk weewx[864] INFO __main__: Using Python 2.7.16 (default, Oct 10 2019, 22:02:15) #012[GCC 8.3.0]
Aug 23 11:57:48 kittyhawk weewx[864] INFO __main__: Platform Linux-5.10.63-v7+-armv7l-with-debian-10.11
Aug 23 11:57:48 kittyhawk weewx[864] INFO __main__: Locale is 'it_IT'

Despite configuration, server reboot, weewx is still using python 2

actually is a not relevant issue since weewx is skipping template, my meteo server is working normally

it is just my curiosity on what can be issue/root cause

thanks for your help!


[2.txt](https://github.com/poblabs/weewx-belchertown/files/9403008/2.txt)
[1.txt](https://github.com/poblabs/weewx-belchertown/files/9403009/1.txt)

**Reply by norecords** [reply]
Look at `/etc/default/weewx` if is set to python 2
```
WEEWX_PYTHON=python2    <== change to python3
WEEWX_BINDIR=/usr/share/weewx
WEEWX_BIN=/usr/bin/weewxd
WEEWX_CFG=/etc/weewx/weewx.conf
```
See that [thread](https://groups.google.com/g/weewx-user/c/vcS60bw-m2U), you may also need to install libs like paho-mqtt for python 3

**Reply by frahugras** [reply]
Hello,

thanks, after setting phyton3 into weewx environment variables, i've rebooted server: no more errors are populating logs: actually i didnt notice i was using old python2; and it was not necessary install mqtt library (at least in my scenario)

really thanks for your help/time

br
Fra

**Reply by norecords** [reply]
Can you set this thread has answered. Thanks you too.

**Reply by frahugras** [reply]
done :)
